### ISSUE TITLE
Spacemacs freezes with basic install

### ISSUE THREAD

**Issue by hsitz** [syl20bnr/spacemacs#4552]
I see some similar issue reports regarding Spacemacs freezing, but I'm not sure whether main issues are fixed or whether they're actually the same as I'm seeing.  I've been experiencing occasional freezes since first starting with Spacemacs a couple months ago.  When it happens it seems to generally be preceded for a second or two by sluggish and/or jerky cursor movement, then complete freeze, need to kill and restart.  It has also happened when I haven't been doing anything at all, e.g., leave computer and come back to find Spacemacs frozen.  These freezes do _not_ result in 100% cpu; they don't seem to affect cpu usage at all.

 I am now seeing these freezes with what is basically a totally standard install of v 0.105.   I'm running Emacs 24.5.1 on Ubuntu with i3wm tiling window manager.  In .spacemacs file my only enabled layers are emacs-lisp and org, otherwise just defaults and these few lines in the user-config function:

  (setq org-default-notes-file "~/notes.org")
  (global-set-key (kbd "C-=") 'text-scale-increase)
  (global-set-key (kbd "C--") 'text-scale-decrease)

Any ideas about what could be causing this?  Anybody else experiencing similar problem?

**Comment by syl20bnr** [comment]
As it happens try to keep a log of what you were doing. As long as we cannot narrow your use cases and cannot reproduce it we won't be able to help much :-(
Also you can provide the output of `SPC h d s`.

**Comment by hsitz** [comment]
On Tue, Jan 12, 2016 at 6:40 AM, Sylvain Benner <notifications@github.com>
wrote:

> As it happens try to keep a log of what you were doing. As long as we
> cannot narrow your use cases and cannot reproduce it we won't be able to
> help much :-(
> Also you can provide the output of SPC h d s.
>
> Thanks, I'll keep watching.  Below is the output of SPC h d s:

#### System Info
- OS: gnu/linux
- Emacs: 24.5.1
- Spacemacs: 0.105.0
- Spacemacs branch: master (rev. 0f7a2e1)
- Distribution: spacemacs
- Layers:
```elisp
(emacs-lisp org)
```

ALSO, my message buffer is littered with the error message below, mostly
preceded by 'byte-code: End of buffer' or 'byte-code: Beginning of buffer'
.   Is this expected?:

Error during redisplay: (eval (spaceline--prepare spaceline-left
spaceline-right)) signaled (void-variable
spacemacs-mode-line-new-version-lighterp)

**Comment by syl20bnr** [comment]
Update your spacemac, the last version is 0.150.8, the spaceline bug is fixed since 0.105.3 ;-)

**Comment by hsitz** [comment]
Thanks, that does fix the spaceline issue.  Was this likely related to my
emacs freezing?

**Comment by syl20bnr** [comment]
I don't think so. Keep us updated about it.

**Comment by anakos** [comment]
Hi -

I am also experiencing Spacemacs freezing on Ubuntu. After a fresh install followed by a restart, spacemacs freezes and is completely unusable.  I am on Ubuntu 15.10 and am running Emacs 24.5.1.  Should I open up a separate issue or just post the relevant information here?

Thanks!
Alex

**Comment by StreakyCobra** [comment]
@anakos Can you try again opening Spacemacs and wait 5 minutes to see if it unfreeze? If this is the case it's probably because of [this](https://github.com/syl20bnr/spacemacs/blob/develop/doc/FAQ.org#why-is-spacemacs-hanging-on-startup). Otherwise, please open a new issue that follows [this template](https://github.com/syl20bnr/spacemacs/wiki/Debugging#how-to-make-a-great-bug-report), thanks!

**Comment by anakos** [comment]
That fixed it - thanks!!!!

**Comment by hsitz** [comment]
I originated this issue, and I think I've found a way to reproduce the problem.  I think it's tied to auto-fill-mode.  If I edit an org document open without auto-fill-mode I don't think I ever get a freeze.  I'm pretty sure I can generate a freeze within a few minutes of activating auto-fill-mode, just cursor around in some lines of text for a while and cursor will eventually stop moving, Emacs frozen.  

I'm not sure cursoring around is even required, because previously I remember coming back to my computer after having left it to find that emacs was frozen.  I haven't experienced any problem at all in last few weeks, and haven't used auto-fill-mode in last few weeks.  Just tried activating auto-fill-mode and now I'm getting the freezes again.  I have generated freezes with message buffer showing, don't see any errors or unexpected messages there.

Using Emacs 24.5.1 and Spacemacs 0.105.8, although I think I've experienced same problem in earlier versions.

**Comment by hsitz** [comment]
I'm original person who submitted this issue, and just want to report that I still have it.  I haven't noticed it much (at all?) over last month or so.  My use of Spacemacs has mostly been limited to opening an org file that I use as reference, i.e., not much editing, just locating portions of it to read.  Yesterday I was actually creating and editing a document, and I started having the freezing problem again.

**It's possible my freezes occur only after heavy use of cursor keys.**  I seem to be able cause the freeze by doing lots of continuous cursor movement for a couple minutes, pressing h,j,k,l,0,$ and/or the arrow keys, maneuvering cursor back and forth over a block of text.

The problem doesn't seem to be limited to a type of file; I've tested with .org file and .tex file today and it happens in both.  

As I've said before, my Spacemacs is a plain-jane install with the org and elisp layers activated.

**Comment by bwinkler** [comment]
What's happening to @hsitz sounds exactly like what's happening with my Spacemacs setup and we have almost identical environments (see below). 

The freezes happen most reliably for me when editing .tex files (that's mostly what I'm editing, though) and I also feel like it happens just when navigating around the file using vim-style movements. That is to say, it doesn't happen when saving or running latexmk or anything like that. 

Also, the CPU is not pegged nor is memory being used up. It just becomes completely unresponsive and the only way out is to kill emacs.

Below is system info output. I'm happy to provide any other info and try anything.

#### System Info
- OS: gnu/linux
- Emacs: 24.5.1
- Spacemacs: 0.105.14
- Spacemacs branch: master (rev. a298a5a)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(auto-completion emacs-lisp markdown org latex themes-megapack deft)
```

**Comment by bb4242** [comment]
I just started using spacemacs (stock config; no customizations) and have been noticing similar issues with intermittent freezes that often seem related to scrolling around using the arrow keys.  I've noticed emacs using 100% of a single CPU core while scrolling.  When I stop scrolling, CPU usage drops back down to very little.

#### System Info
- OS: gnu/linux
- Emacs: 24.5.1
- Spacemacs: 0.105.19
- Spacemacs branch: master (rev. 3274e7d)
- Graphic display: t
- Distribution: spacemacs
- Editing style: emacs
- Completion: helm
- Layers:
```elisp
nil```

**Comment by bb4242** [comment]
Just dug into this a bit more, and ran a cpu and memory [profile](http://www.gnu.org/software/emacs/manual/html_node/elisp/Profiling.html) in spacemacs while scrolling from the top to the bottom of a short 250-line file (my .spacemacs config).  I'm not an expert on emacs profiling by any means, but the memory profile in particular seemed suspicious:

```
- command-execute                                         674,700,892  84%
 - call-interactively                                     674,690,492  84%
  - next-line                                             639,112,330  79%
   - apply                                                639,112,330  79%
    - ad-Advice-next-line                                 639,112,330  79%
     - #<compiled 0x24d4f9>                               518,512,638  64%
      - line-move                                         518,449,198  64%
       - line-move-partial                                514,422,935  64%
        - pos-visible-in-window-p                         333,402,540  41%
         - eval                                           333,402,540  41%
          - spaceline--prepare                            330,758,088  41%
           - run-hooks                                    272,506,596  34%
            - spacemacs//prepare-diminish                 272,506,596  34%
             - if                                         272,506,596  34%
              - progn                                     272,506,596  34%
               - let                                      272,506,596  34%
                - let                                     272,506,596  34%
                 - while                                  267,819,298  33%
                  - let                                   267,817,680  33%
                   - if                                   267,817,680  33%
                    - progn                               267,817,680  33%
                     - let*                               267,817,680  33%
                        diminish                          267,771,920  33%
           - spaceline--prepare-any                        53,299,004   6%
            - spaceline--eval-segment                      41,983,244   5%
             - mapcar                                      22,187,140   2%
              - #<compiled 0x15f863f>                      21,141,788   2%
               - spaceline--eval-segment                   21,049,228   2%
                + mapcar                                    4,080,140   0%
                + powerline-raw                             2,718,152   0%
                + spaceline--segment-minor-modes            1,895,120   0%
                + spaceline--segment-buffer-id              1,423,196   0%
                + eval                                      1,009,840   0%
                + spaceline--segment-buffer-size              359,884   0%
                  spaceline--segment-buffer-encoding-abbrev        49,128   0%
              + #<compiled 0x15f8653>                       1,045,352   0%
             + powerline-raw                                4,320,320   0%
             + spaceline--segment-hud                       1,656,508   0%
             + spaceline--eval-segment                        727,296   0%
             + spaceline--segment-major-mode                  535,824   0%
             + eval                                           241,900   0%
             + cl-remove-if-not                               204,880   0%
             + spaceline--intersperse                          48,880   0%
               apply                                           34,320   0%
             + spaceline--segment-new-version                  34,320   0%
            + mapcar                                        7,340,124   0%
            + -zip                                            274,560   0%
              apply                                           154,960   0%
           + powerline-render                               3,082,052   0%
           + powerline-width                                1,703,620   0%
             powerline-fill                                    72,176   0%
        + posn-at-point                                   165,086,859  20%
        + window-screen-lines                                 811,448   0%
        + default-line-height                                 739,440   0%
          default-font-height                                 314,988   0%
       + default-line-height                                  631,016   0%
       + line-move-visual                                     548,932   0%
      + called-interactively-p                                 42,640   0%
     + do-smooth-scroll                                   120,593,452  15%
  + byte-code                                              29,181,653   3%
  + helm-M-x                                                6,393,389   0%
+ redisplay_internal (C function)                         124,969,470  15%
+ hl-paren-initiate-highlight                                  90,464   0%
+ sp--save-pre-command-state                                   88,400   0%
+ timer-event-handler                                          63,248   0%
+ winner-save-old-configurations                               54,064   0%
+ which-key--hide-popup                                         5,200   0%
  ...                                                               0   0%
```
If I'm reading this correctly, it means that 670MB of memory was allocated (and maybe garbage collected) in the course of scrolling from the top to the bottom of the file.

Here's the corresponding CPU profile:

```
- command-execute                                                7255  88%
 - call-interactively                                            7255  88%
  - next-line                                                    5669  69%
   - apply                                                       5669  69%
    - ad-Advice-next-line                                        5665  68%
     - #<compiled 0x24d4f9>                                      4114  50%
      - line-move                                                4114  50%
       - line-move-partial                                       4002  48%
        - pos-visible-in-window-p                                1509  18%
         - eval                                                  1509  18%
          - spaceline--prepare                                   1505  18%
           - spaceline--prepare-any                              1012  12%
            - spaceline--eval-segment                             874  10%
             - mapcar                                             435   5%
              - #<compiled 0x15f863f>                             419   5%
               - spaceline--eval-segment                          391   4%
                + spaceline--segment-minor-modes                   73   0%
                + powerline-raw                                    45   0%
                + mapcar                                           35   0%
                  facep                                            25   0%
                + spaceline--segment-buffer-size                   11   0%
                  spaceline--segment-buffer-encoding-abbrev                  8   0%
                + eval                                              8   0%
                + spaceline--segment-window-number                  5   0%
                + spaceline--segment-buffer-id                      5   0%
              + #<compiled 0x15f8653>                               8   0%
             + powerline-raw                                       89   1%
             + spaceline--segment-hud                              36   0%
               facep                                               34   0%
             + spaceline--eval-segment                             23   0%
             + eval                                                15   0%
               spaceline--segment-mu4e-alert-segment                 14   0%
             + spaceline--segment-major-mode                        8   0%
             + spaceline--segment-global                            4   0%
            + mapcar                                               63   0%
            + -zip                                                  9   0%
           + run-hooks                                            447   5%
           + powerline-width                                       26   0%
           + powerline-render                                       8   0%
        + posn-at-point                                           735   8%
          default-font-height                                       8   0%
        + window-screen-lines                                       6   0%
        + default-line-height                                       3   0%
       + line-move-visual                                          96   1%
     + do-smooth-scroll                                          1551  18%
  + byte-code                                                    1458  17%
  + previous-line                                                  68   0%
  + helm-M-x                                                       55   0%
  + profiler-report-toggle-entry                                    5   0%
+ redisplay_internal (C function)                                 566   6%
+ ...                                                             352   4%
+ global-hl-line-highlight                                         17   0%
+ timer-event-handler                                              13   0%
+ sp--save-pre-command-state                                        6   0%
+ winner-save-old-configurations                                    4   0%
+ which-key--hide-popup                                             1   0%
```

**Comment by NJBS** [comment]
@bb4242 Can you reproduce if you turn off smooth-scrolling?

**Comment by syl20bnr** [comment]
I was going to ask the same question as @NJBS, also can you try with spaceline excluded ?

**Comment by bb4242** [comment]
OK, I tried both turning off smooth scrolling (by setting `dotspacemacs-smooth-scrolling nil` in `~/.spacemacs`) and spaceline (by commenting out `spaceline` from the `(setq spacemacs-packages...)` section of `layers/+distribution/spacemacs/packages.el` - not sure if that's the right way to do it, but it did seem to turn it off).

Turning off smooth scrolling has no effect.

Turning off spaceline makes things much better - CPU usage is much lower while scrolling, the scrolling is much smoother, and I don't observe the kind of freezing behavior I was seeing with spaceline turned on.

**Comment by StreakyCobra** [comment]
CC @TheBB

**Comment by TheBB** [comment]
Well, it's no secret that spaceline isn't the best at performance, though I've never had problems with it, scrolling or not.

The memory profile seems to point to `spacemacs//prepare-diminish`, which indeed runs much more often than it really needs to. Does the following workaround help?

```elisp
(remove-hook 'spaceline-pre-hook 'spacemacs//prepare-diminish)
```

Otherwise can you try removing segments from spaceline to see if any of them are responsible? I don't think they should be (the actual preparation function is probably more expensive) but it's worth a go. This involves removing stuff from `spaceline-left` and `spaceline-right`.

**Comment by syl20bnr** [comment]
FWIW `smooth-scrolling` is not an Emacs default, by default the buffer scrolls by chunks. While it seems ugly at first because we are used to see the content scrolling line by line in other apps, I found that after a few days I got used to the default jumping behaviour and found that it is a lot better than `smooth-scrolling` because it lowers eye strain by a huge amount for me. To scroll the contents on request there is `C-y` and `C-e` also `SpC z f` to display the whole function on the screen if possible.
Now the only situation where I would use `smooth-scrolling` is for presentation purpose.

**Comment by JAremko** [comment]
Also it looks like `jit-lock-register` runs functions(without hacks) only when the part of previously invisible region was scrolled into visible area with decent margin. I think it causes visual popping of fontifications(in my code) Also `smooth-scrolling` appears to be not so smooth due to all the hookery and font lock  shenanigans.

**Comment by bb4242** [comment]
@TheBB : The `(remove-hook 'spaceline-pre-hook 'spacemacs//prepare-diminish)` workaround doesn't seem to have any observable effect.  I also tried setting both `spaceline-left` and `spaceline-right` to `nil`.  When both are set to `nil` simultaneously, the freezes go away.  With only one of them set to `nil`, the freezes still occur (although they might be a little less frequent, it's hard to tell).  So, it appears that at there are at least two segments that contribute to the freezes (one in `spaceline-left` and one in `spaceline-right`).

Thanks for your help, everyone - spacemacs is really cool! :)

**Comment by ctjhoa** [comment]
I also have this issue. I've some profiling data from the hang (100% CPU) time.

![2016-04-28-182410_1025x610_scrot](https://cloud.githubusercontent.com/assets/1716173/14893269/f4b7ccc8-0d6e-11e6-841c-5acacec7d56f.png)

**Comment by syl20bnr** [comment]
To people having the issue, do you have a higher than usual key repeat frequency ?

**Comment by bb4242** [comment]
Just checked - my key repeat frequency is 30 repeats/sec.

**Comment by syl20bnr** [comment]
Mine is at 40 and I don't encounter the issue, so we can forget this.

One thing we could try also is to replace spaceline by powerline directly, to do so:
- add `spaceline` to `dotspacemacs-excluded-packages`
- add `powerline` to `dotspacemacs-additonal-packages`
- in your `user-config` function add:

```elisp
(require 'powerline)
(powerline-default-theme)
```

Also can you all provide the following information ?
- used font
- used theme

**Comment by bb4242** [comment]
Replacing spaceline with powerline makes things much smoother for me - thanks for that suggestion!

I'm using the default font configuration shipped with spacemacs; here's the relevant .spacemacs section:
```
   dotspacemacs-default-font '("Source Code Pro" 
                               :size 13
                               :weight normal
                               :width normal
                               :powerline-scale 1.1)
```

I've generally been using the `spacemacs-dark` theme, although I've tried a couple different ones and noticed the freezes with all of them.

**Comment by ctjhoa** [comment]
My font :
```
   ;; size to make separators look not too crappy.
   dotspacemacs-default-font '("DejaVu Sans Mono"
                               :size 18
                               :weight normal
                               :width normal
                               :powerline-scale 1.1)
```
And I use both themes monokai and spacemacs-dark.

**Comment by ctjhoa** [comment]
Another thing to notice, I work mostly with split windows (often 6 windows or more)

**Comment by izahn** [comment]
I have this problem too. In my desktop machine I don't notice it as much, but on my old slow laptop it is unusable. I'm using the "replace spaceline with powerline" workaround for now.

Here is my CPU profile in case it helps (looks pretty much the same as what others have already posted).

```
- command-execute                                               14480  94%
 - call-interactively                                           14477  94%
  - next-line                                                   13646  89%
   - apply                                                      13646  89%
    - ad-Advice-next-line                                       13646  89%
     - #<compiled 0x24d09f>                                     13646  89%
      - line-move                                               13643  89%
       - line-move-partial                                      13398  87%
        - pos-visible-in-window-p                                4069  26%
         - eval                                                  4069  26%
          - spaceline--prepare                                   4053  26%
           - spaceline--prepare-any                              3905  25%
            - spaceline--eval-segment                            3201  20%
             - mapcar                                            1645  10%
              - #<compiled 0x19bf587>                            1595  10%
               + spaceline--eval-segment                         1518   9%
              + #<compiled 0x19bf59b>                              34   0%
```

**Comment by TheBB** [comment]
I have an idea for maybe improving performance, but it's going to be somewhat involved. :-/

**Comment by TheBB** [comment]
The `codegen` branch of Spaceline is now mostly functional. Some minor features remain + cleanup and documentation.

It's hard to get stable profiling results but I'm looking at roughly the following numbers (all of them *per modeline evaluation*)

 | Compiled | CPU samples | kB allocated |
---|---|---|---
**Master** | | 1.63–1.65 | 163–173
**Codegen** | no | 0.745–0.769 | 92.8–93.8
**Codegen** | yes | 0.294–0.299 | 75.9–76.2

The "compiled" column refers to whether the modeline evaluation function (which is generated on the fly according to a modeline spec) is also byte-compiled or not. This will be a user option. I'm not sure if it'll be on or off by default, I'll have to see if it causes any troubles.

**Comment by TheBB** [comment]
By the way, this is the profiling script I've been using.

```elisp
(dotimes (_ 1000)
  (spaceline--prepare spaceline-left spaceline-right))
(profiler-start 'cpu+mem)
(dotimes (_ 10000)
  (spaceline--prepare spaceline-left spaceline-right))
(profiler-report)
(profiler-stop)
```

If you want to test out the codegen branch you have to replace the modeline eval function with `(spaceline-ml-main)`.

**Comment by TheBB** [comment]
A new and faster version of Spaceline is on MELPA now.

**Comment by bb4242** [comment]
Just updated to the new spaceline version, and things seem much smoother now.  I'll continue testing, but my initial impression is that this new version will make it possible for me to completely switch back to spaceline (I had needed to switch to powerline to make spacemacs usable for me).  Thanks, @TheBB ! :)

**Comment by hsitz** [comment]
I've updated to what I think are newest versions -- spacemacs (0.105.20 ) and  spaceline (2016510.1330) -- and I still have the freezing problem just as much as I ever did.  Possibly more.  I have a file I test this with, a simple Org outline, 600 lines of text, about 28k.  If I open it, move to a column somewhere in middle of the text, and hold the down arrow key to cursor down through the buffer it consistently freezes before I reach the bottom.

I have this problem only in gui Emacs, not when using Emacs in terminal.  I do see the occasional slowdown/sputtering of cursor in terminal version, but I can't recall it ever freezing on me.

I don't run on high powered machine, but it's not dog slow either, it's laptop with CoreM 5Y10 processor, 8GB RAM, 256 GB SSD.

**Comment by aaronjensen** [comment]
@hsitz a profile for 30-60 seconds of usage with `profiler-report-render-reversed-calltree` would be helpful.

**Comment by hsitz** [comment]
I don't seem to have any improvement with this issue, even from recent updates.  Not sure, scrolling may be slightly zippier.  But the freezes, if anything, are more frequent and easier to induce.  

I have a test org file (attached) that I use to induce a freeze.   I open the file, do <shift-tab> (Org's cycle all headings) to make sure all headings and text is displayed, cursor to middle of first line of body text, then hold down arrow key to cursor to bottom of document.  It consistently freezes at between 45% to 55% to end of document.

I have attached profile reports that I obtained after using procedure above, but stopping at 40% down, i.e., a bit before I would expect a freeze.  Even so, I had issues generating and processing the report text.  When editing the report text (in same spacemacs instance as it was generated) I would sometimes get a freeze.  This happened many minutes after cursoring down test document, and it wasn't tied to cursor key usage.  (E.g., once it happened while I was holding '=' key to generate a line of spacing between reports.)  IN general, this was similar to what I've experienced before: at times I've been working in a document, left it for an hour or more, then have freeze happen almost immediately upon coming back to it, without many keystrokes at all.  (I finally ended up just saving the profiles and reloading them in a terminal instance of spacemacs before editing.)

The profiles attached were generated using Spacemacs 0.105.20 and with updated packages.  I just tested with 0.105.21 and behavior was the same.
[test.zip](https://github.com/syl20bnr/spacemacs/files/310412/test.zip)
[bug-profiles.txt](https://github.com/syl20bnr/spacemacs/files/310413/bug-profiles.txt)

**Comment by quicknir** [comment]
I'm experience this as well. This only happens when scrolling down, not up, and only in insert mode (or emacs mode), not in normal mode. The relevant portion of my profile report:

```
- command-execute                                               10030  93%
 - call-interactively                                           10030  93%
  - next-line                                                    8955  83%
   - apply                                                       8955  83%
    - ad-Advice-next-line                                        8955  83%
     - #<compiled 0x26a477>                                      8955  83%
      - line-move                                                8953  83%
       - line-move-partial                                       8493  78%
        + pos-visible-in-window-p                                 650   6%
        + posn-at-point                                           317   2%
        + default-line-height                                      13   0%
        + window-screen-lines                                       9   0%
       + line-move-visual                                         412   3%
       + default-line-height                                       19   0%
      + called-interactively-p                                      1   0%
  - byte-code                                                    1040   9%
   + helm-M-x-read-extended-command                              1040   9%
  + evil-normal-state                                              28   0%
  + helm-M-x                                                        5   0%
  + evil-insert                                
```

**Comment by ACOOD** [comment]
I'm coming across a problem which is very similar to @quicknir 's. It can scroll up more faster than scroll down. And the profile report is also similar to @quicknir 's report. It will not happen in terminal.

I had tried all methods in this issue, but they didn't work. And it behave different in different files, it is easy to freeze in .tex files.

I'm using the newest spacemacs(0.200.7) and emacs(25.1.1 64bit) in Win10.

**Comment by syl20bnr** [comment]
What font are you using in Emacs ?

**Comment by ACOOD** [comment]
I'm using the default font.
"Source Code Pro"
:size 13
:weight normal
:width normal
:powerline-scale 1.1

**Comment by dhanjit** [comment]
Is there any update for this. I am seeing lags while navigation, moving cursor basically. Running a top on a separate external terminal gives me 90+% cpu usage on simple navigation. I binded "u" to profiler-report to quickly show profiler output so that it doesn't get affect by the helm-M-x. The culprit is clearly spaceline. I changed config to use vim-powerline/vanilla/doom modes for the modeline and the lag immediately decreased after restarting emacs. Clearly it's a issue with the default spaceline.

Latest develop branch spacemacs and emacs 26.3
